Async & Aztec: Removing a video from a post could lead to crash

It seems that removing a video from a post in Aztec does't really stop the uploading of the video, instead I see my connection busy (sending the video I guess) and other pictures are postponed until the video network upload is done.
At the end of the uploading process for the video the app crashed with the following error:

```
08-22 09:12:55.809 22585-22585/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
                                                                       Process: org.wordpress.android, PID: 22585
                                                                       org.greenrobot.eventbus.EventBusException: Invoking subscriber failed
                                                                           at org.greenrobot.eventbus.EventBus.handleSubscriberException(EventBus.java:505)
                                                                           at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:487)
                                                                           at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479)
                                                                           at org.greenrobot.eventbus.HandlerPoster.handleMessage(HandlerPoster.java:67)
                                                                           at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                           at android.os.Looper.loop(Looper.java:154)
                                                                           at android.app.ActivityThread.main(ActivityThread.java:6121)
                                                                           at java.lang.reflect.Method.invoke(Native Method)
                                                                           at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
                                                                           at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
                                                                        Caused by: java.lang.IllegalArgumentException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkParameterIsNotNull, parameter attributes
                                                                           at org.wordpress.aztec.AztecAttributes.<init>(AztecAttributes.kt)
                                                                           at org.wordpress.android.editor.MetadataUtils$AttributesWithClass.<init>(MetadataUtils.java:131)
                                                                           at org.wordpress.android.editor.AztecEditorFragment.markMediaFailed(AztecEditorFragment.java:1562)
                                                                           at org.wordpress.android.ui.uploads.MediaUploadReadyProcessor.markMediaUploadFailedInPost(MediaUploadReadyProcessor.java:43)
                                                                           at org.wordpress.android.ui.uploads.UploadService.updatePostWithFailedMedia(UploadService.java:520)
                                                                           at org.wordpress.android.ui.uploads.UploadService.updatePostWithCurrentlyFailedUploads(UploadService.java:373)
                                                                           at org.wordpress.android.ui.uploads.UploadService.updatePostModelWithCompletedAndFailedUploads(UploadService.java:569)
                                                                           at org.wordpress.android.ui.uploads.UploadService.stopServiceIfUploadsComplete(UploadService.java:549)
                                                                           at org.wordpress.android.ui.uploads.UploadService.onMediaUploaded(UploadService.java:727)
                                                                           at java.lang.reflect.Method.invoke(Native Method)
                                                                           at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
                                                                           at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479) 
                                                                           at org.greenrobot.eventbus.HandlerPoster.handleMessage(HandlerPoster.java:67) 
                                                                           at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                           at android.os.Looper.loop(Looper.java:154) 
                                                                           at android.app.ActivityThread.main(ActivityThread.java:6121) 
                                                                           at java.lang.reflect.Method.invoke(Native Method) 
                                                                           at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889) 
                                                                           at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779) 
```

Video here: https://cloudup.com/cbns9SNSTW7

I also noticed (and you can see in the video) that Media Library does contain multiple instances of the video. Those items cannot be open, as shown in the video.


I experienced this issue in `develop`, but I guess it's applicable to `8.1` too. Should we fix this in 8.1 instead?

--- a/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
+++ b/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
@@ -876,6 +876,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
     @Override
     public void onMediaDeleted(AztecAttributes aztecAttributes) {
         String localMediaId = aztecAttributes.getValue(ATTR_ID_WP);
+        mUploadingMediaProgressMax.remove(localMediaId);
         if (!TextUtils.isEmpty(localMediaId)) {
             mEditorFragmentListener.onMediaDeleted(localMediaId);
         }

--- a/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
+++ b/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
@@ -1655,13 +1655,13 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
     }
 
     private static void resetMediaWithStatus(Spanned content, String status) {
-        // get all items with "uploading" class
-        AztecText.AttributePredicate uploadingPredicate = getPredicateWithClass(status);
+        // get all items with class defined by the "status" variable
+        AztecText.AttributePredicate statusPredicate = getPredicateWithClass(status);
 
         // update all items to failed, unless they already have a remote URL, in which case
         // it means the upload completed, but the item remained inconsistently marked as uploading
         // (for example after an app crash)
-        for (IAztecAttributedSpan span : getSpansForPredicate(content, uploadingPredicate, false)) {
+        for (IAztecAttributedSpan span : getSpansForPredicate(content, statusPredicate, false)) {
             clearMediaUploadingAndSetToFailedIfLocal(span);
         }
     }

--- a/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
+++ b/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
@@ -545,12 +545,16 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
         AztecText.AttributePredicate predicate = new AztecText.AttributePredicate() {
             @Override
             public boolean matches(@NonNull Attributes attrs) {
-                AttributesWithClass attributesWithClass = new AttributesWithClass(attrs);
+                AttributesWithClass attributesWithClass = getAttributesWithClass(attrs);
                 return attributesWithClass.hasClass(classToUse);
             }
         };
 
         return predicate;
+    }
+
+    static private AttributesWithClass getAttributesWithClass(@NonNull Attributes attrs) {
+        return new AttributesWithClass(attrs);
     }
 
     private void updateFailedMediaList() {
@@ -570,7 +574,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
             overlayProgressingMedia(predicate);
             // here check if this is a video uploading in progress or not; if it is, show the video play icon
             for (Attributes attrs : content.getAllElementAttributes(predicate)) {
-                AttributesWithClass attributesWithClass = new AttributesWithClass(attrs);
+                AttributesWithClass attributesWithClass = getAttributesWithClass(attrs);
                 if (attributesWithClass.hasClass(TEMP_VIDEO_UPLOADING_CLASS)) {
                     overlayVideoIcon(2, predicate);
                 }
@@ -798,7 +802,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
         content.removeMedia(new AztecText.AttributePredicate() {
             @Override
             public boolean matches(@NotNull Attributes attrs) {
-                return new AttributesWithClass(attrs).hasClass(ATTR_STATUS_FAILED);
+                return getAttributesWithClass(attrs).hasClass(ATTR_STATUS_FAILED);
             }
         });
         mFailedMediaIds.clear();
@@ -839,7 +843,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
                 MediaPredicate predicate = MediaPredicate.getLocalMediaIdPredicate(localMediaId);
 
                 // remove the uploading class
-                AttributesWithClass attributesWithClass = new AttributesWithClass(
+                AttributesWithClass attributesWithClass = getAttributesWithClass(
                         content.getElementAttributes(predicate));
                 attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
                 if (mediaFile.isVideo()) {
@@ -959,7 +963,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
                 case IMAGE:
                 case VIDEO:
                     MediaPredicate localMediaIdPredicate = MediaPredicate.getLocalMediaIdPredicate(localMediaId);
-                    AttributesWithClass attributesWithClass = new AttributesWithClass(
+                    AttributesWithClass attributesWithClass = getAttributesWithClass(
                             content.getElementAttributes(localMediaIdPredicate));
 
                     attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
@@ -1177,7 +1181,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
         Set<String> classes = MetadataUtils.getClassAttribute(attrs);
         String idName;
         String uploadStatus = "";
-        final JSONObject meta = MetadataUtils.getMetadata(new AttributesWithClass(attrs), naturalWidth, naturalHeight);
+        final JSONObject meta = MetadataUtils.getMetadata(getAttributesWithClass(attrs), naturalWidth, naturalHeight);
         if (classes.contains(ATTR_STATUS_UPLOADING)) {
             uploadStatus = ATTR_STATUS_UPLOADING;
             idName = ATTR_ID_WP;
@@ -1244,7 +1248,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
                 switch (mediaType) {
                     case IMAGE:
                     case VIDEO:
-                        AttributesWithClass attributesWithClass = new AttributesWithClass(
+                        AttributesWithClass attributesWithClass = getAttributesWithClass(
                                 content.getElementAttributes(mTappedMediaPredicate));
 
                         // remove the failed class
@@ -1393,7 +1397,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
                     attributes.setValue(ATTR_ALT, JSONUtils.getString(meta, ATTR_ALT));
                 }
 
-                AttributesWithClass attributesWithClass = new AttributesWithClass(attributes);
+                AttributesWithClass attributesWithClass = getAttributesWithClass(attributes);
 
                 // remove previously set class attributes to add updated values
                 attributesWithClass.removeClassStartingWith(ATTR_ALIGN_DASH);
@@ -1456,7 +1460,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
     }
 
     private static void addDefaultSizeClassIfMissing(AztecAttributes attributes) {
-        AttributesWithClass attrs = new AttributesWithClass(attributes);
+        AttributesWithClass attrs = getAttributesWithClass(attributes);
         if (!attrs.hasClassStartingWith("size")) {
             attrs.addClass("size-full");
         }
@@ -1466,7 +1470,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
     // this is used for reattachment: when the editor is opened again on a Post that has in-progress
     // video uploads, we need to show the progress bar and the video play icon to differentiate from images
     private static void addVideoUploadingClassIfMissing(AztecAttributes attributes) {
-        AttributesWithClass attrs = new AttributesWithClass(attributes);
+        AttributesWithClass attrs = getAttributesWithClass(attributes);
         if (!attrs.hasClass(TEMP_VIDEO_UPLOADING_CLASS)) {
             attrs.addClass(TEMP_VIDEO_UPLOADING_CLASS);
         }
@@ -1539,23 +1543,27 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
             MediaPredicate predicate = MediaPredicate.getLocalMediaIdPredicate(localMediaId);
 
             // remove the uploading class
-            AttributesWithClass attributesWithClass = new AttributesWithClass(
-                    getFirstElementAttributes(content, predicate));
-            attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
-            if (mediaFile.isVideo()) {
-                attributesWithClass.removeClass(TEMP_VIDEO_UPLOADING_CLASS);
+            Attributes firstElementAttributes = getFirstElementAttributes(content, predicate);
+            // let's make sure the element is still there within the content. Sometimes it may happen
+            // this method is called but the element doesn't exist in the post content anymore
+            if (firstElementAttributes != null) {
+                AttributesWithClass attributesWithClass = getAttributesWithClass(firstElementAttributes);
+                attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
+                if (mediaFile.isVideo()) {
+                    attributesWithClass.removeClass(TEMP_VIDEO_UPLOADING_CLASS);
+                }
+
+                // add then new src property with the remoteUrl
+                AztecAttributes attrs = attributesWithClass.getAttributes();
+                attrs.setValue("src", remoteUrl);
+
+                addDefaultSizeClassIfMissing(attrs);
+
+                updateElementAttributes(content, predicate, attrs);
+
+                // re-set the post content
+                postContent = parser.toHtml(content, false);
             }
-
-            // add then new src property with the remoteUrl
-            AztecAttributes attrs = attributesWithClass.getAttributes();
-            attrs.setValue("src", remoteUrl);
-
-            addDefaultSizeClassIfMissing(attrs);
-
-            updateElementAttributes(content, predicate, attrs);
-
-            // re-set the post content
-            postContent = parser.toHtml(content, false);
         }
         return postContent;
     }
@@ -1570,20 +1578,25 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
             MediaPredicate predicate = MediaPredicate.getLocalMediaIdPredicate(localMediaId);
 
             // remove the uploading class
-            AttributesWithClass attributesWithClass = new AttributesWithClass(
-                    getFirstElementAttributes(content, predicate));
-            attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
-            if (mediaFile.isVideo()) {
-                attributesWithClass.removeClass(TEMP_VIDEO_UPLOADING_CLASS);
+            Attributes firstElementAttributes = getFirstElementAttributes(content, predicate);
+            // let's make sure the element is still there within the content. Sometimes it may happen
+            // this method is called but the element doesn't exist in the post content anymore
+            if (firstElementAttributes != null) {
+                AttributesWithClass attributesWithClass = getAttributesWithClass(
+                        firstElementAttributes);
+                attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
+                if (mediaFile.isVideo()) {
+                    attributesWithClass.removeClass(TEMP_VIDEO_UPLOADING_CLASS);
+                }
+
+                // mark failed
+                attributesWithClass.addClass(ATTR_STATUS_FAILED);
+
+                updateElementAttributes(content, predicate, attributesWithClass.getAttributes());
+
+                // re-set the post content
+                postContent = parser.toHtml(content, false);
             }
-
-            // mark failed
-            attributesWithClass.addClass(ATTR_STATUS_FAILED);
-
-            updateElementAttributes(content, predicate, attributesWithClass.getAttributes());
-
-            // re-set the post content
-            postContent = parser.toHtml(content, false);
         }
         return postContent;
     }
@@ -1668,8 +1681,7 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
 
     private static void clearMediaUploadingAndSetToFailedIfLocal(IAztecAttributedSpan span) {
         // remove the uploading class
-        AttributesWithClass attributesWithClass = new AttributesWithClass(
-                span.getAttributes());
+        AttributesWithClass attributesWithClass = getAttributesWithClass(span.getAttributes());
         attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
 
         attributesWithClass = addFailedStatusToMediaIfLocalSrcPresent(attributesWithClass);

--- a/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
+++ b/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java
@@ -1501,6 +1501,20 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
         return allAttrs;
     }
 
+    private static @NonNull List<IAztecAttributedSpan> getSpansForPredicate(Spanned content,
+                                                                            AztecText.AttributePredicate predicate,
+                                                                            boolean returnFirstFoundOnly) {
+        IAztecAttributedSpan[] spans = content.getSpans(0, content.length(), IAztecAttributedSpan.class);
+        List<IAztecAttributedSpan> allMatchingSpans = new ArrayList<>();
+        for (IAztecAttributedSpan span : spans) {
+            if (predicate.matches(span.getAttributes())) {
+                allMatchingSpans.add(span);
+                if (returnFirstFoundOnly) return allMatchingSpans;
+            }
+        }
+        return allMatchingSpans;
+    }
+
     private static void updateElementAttributes(Spanned content,
                                                 AztecText.AttributePredicate predicate,
                                                 AztecAttributes attrs) {
@@ -1627,7 +1641,12 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
 
     public void setMediaToFailed(@NonNull String mediaId) {
         AztecText.AttributePredicate localMediaIdPredicate = MediaPredicate.getLocalMediaIdPredicate(mediaId);
-        clearMediaUploadingAndSetToFailedIfLocal(content.getText(), localMediaIdPredicate);
+        // we should be obtaining just one span for this media Id predicate, but just in case something
+        // weird happened we make sure we run through all obtained spans and mark them failed if a local src found
+        List<IAztecAttributedSpan> spans = getSpansForPredicate(content.getText(), localMediaIdPredicate, false);
+        for (IAztecAttributedSpan span : spans) {
+            clearMediaUploadingAndSetToFailedIfLocal(span);
+        }
         content.clearOverlays(localMediaIdPredicate);
         overlayFailedMedia(mediaId, content.getElementAttributes(localMediaIdPredicate));
         safeAddMediaIdToSet(mFailedMediaIds, mediaId);
@@ -1641,36 +1660,21 @@ public class AztecEditorFragment extends EditorFragmentAbstract implements
         // update all items to failed, unless they already have a remote URL, in which case
         // it means the upload completed, but the item remained inconsistently marked as uploading
         // (for example after an app crash)
-        for (Attributes attrs : getAllElementAttributes(content, uploadingPredicate)) {
-            clearMediaUploadingAndSetToFailedIfLocal(content, getPredicateForMedia(attrs));
+        for (IAztecAttributedSpan span : getSpansForPredicate(content, uploadingPredicate, false)) {
+            clearMediaUploadingAndSetToFailedIfLocal(span);
         }
     }
 
-    private static AztecText.AttributePredicate getPredicateForMedia(Attributes attrs) {
-        String itemId = attrs.getValue(ATTR_ID_WP);
-        AztecText.AttributePredicate predicate;
-
-        if (!TextUtils.isEmpty(itemId)) {
-            predicate = MediaPredicate.getLocalMediaIdPredicate(itemId);
-        } else {
-            // if ATTR_ID_WP is missing, try with TEMP_IMAGE_ID
-            itemId = attrs.getValue(TEMP_IMAGE_ID);
-            predicate = MediaPredicate.getTempMediaIdPredicate(itemId);
-        }
-        return predicate;
-    }
-
-    private static void clearMediaUploadingAndSetToFailedIfLocal(Spanned content, AztecText.AttributePredicate predicate) {
+    private static void clearMediaUploadingAndSetToFailedIfLocal(IAztecAttributedSpan span) {
         // remove the uploading class
         AttributesWithClass attributesWithClass = new AttributesWithClass(
-                getFirstElementAttributes(content,predicate));
+                span.getAttributes());
         attributesWithClass.removeClass(ATTR_STATUS_UPLOADING);
 
         attributesWithClass = addFailedStatusToMediaIfLocalSrcPresent(attributesWithClass);
 
-        updateElementAttributes(content, predicate, attributesWithClass.getAttributes());
+        span.setAttributes(attributesWithClass.getAttributes());
     }
-
 
     private static AttributesWithClass addFailedStatusToMediaIfLocalSrcPresent(AttributesWithClass attributesWithClass) {
         // check if "src" value is remote or local, it only makes sense to mark failed local files

